### PR TITLE
bool filter, add booleanization strategy option

### PR DESCRIPTION
**Pull request by bcoca**
also make the YAML booleanization the same as spec/JSON/module function
 previous 'aproximation' was missing several options.

##### ISSUE TYPE

- Feature Pull Request

### CODE REVIEW

File: lib/ansible/plugins/filter/bool.yml [hunk 1]
```diff
@@ -10,17 +10,31 @@ DOCUMENTATION:
       description: Data to cast.
       type: raw
       required: true
+    strategy:
+      description: Which evaluation engine to use
```
Comment (Akasurde) [inline lib/ansible/plugins/filter/bool.yml]:
```suggestion
      description: Evaluation engine to use while evaluating boolean.
```

File: lib/ansible/plugins/filter/bool.yml [hunk 2]
```diff
@@ -10,17 +10,31 @@ DOCUMENTATION:
...
+      default: yaml
+      choices:
+        yaml: Use YAML 1.1 allowed booleans
+        python: Pass through the Python C(bool) function
+        truthy: Just let Python evaluate in in an C(if) statement
```
Comment (Akasurde) [inline lib/ansible/plugins/filter/bool.yml]:
```suggestion
        truthy: Just let Python evaluate in an C(if) statement
```

File: lib/ansible/plugins/filter/bool.yml [hunk 3]
```diff
@@ -10,17 +10,31 @@ DOCUMENTATION:
...
+      version_added: '2.18'
+  notes:
+    - prior to adding the 'strategy' option, evaluation was aproximating the 'yaml' strategy.
+    - if the value is already a boolean or C(None) this filter just returns the original value.
```
Comment (Akasurde) [inline lib/ansible/plugins/filter/bool.yml]:
```suggestion
    - Before adding the 'strategy' option, evaluation approximated the 'yaml' strategy.
    - if the value is already a boolean or C(None), this filter returns the original value.
```

File: lib/ansible/plugins/filter/core.py [hunk 4]
```diff
@@ -80,14 +81,27 @@ def to_nice_json(a, indent=4, sort_keys=True, *args, **kw):
...
+        elif strategy == 'python':
+            return bool(a)
+        elif strategy == 'truthy':
+            if a:
+                return True
+        else:
+            raise AnsibleFilterError(f"Invalid strategy provided ({strategy}), valid choices are: yaml, python, truthy")
+    except TypeError as e:
+        raise AnsibleFilterTypeError(f"Could not convert to boolean: {e}", orig_exc=e)
+
```
Comment (Akasurde) [inline lib/ansible/plugins/filter/core.py]:
```suggestion
    valid_choices = ('yaml', 'python', 'truthy')
    if strategy not in valid_choices:
        raise AnsibleFilterError(f"Invalid strategy provided ({strategy}), valid choices are: {', '.join(valid_choices)}")

    try:
        if strategy == 'yaml':
            # make it lower case for easier matching
            return boolean(a)
        if strategy == 'python':
            return bool(a)
        if strategy == 'truthy':
            if a:
                return True
    except TypeError as e:
        raise AnsibleFilterTypeError(f"Could not convert to boolean using strategy {strategy}: {e}", orig_exc=e)
```

**Review by Akasurde — APPROVED** [review]

**Comment by webknjaz** [comment]
@mattclay looks like the change detection in `ansible-test` didn't run the right tests? What does need to be updated to make it make the proper connection?

I see that @s-hertel already has a PR to fix the broken tests themselves: https://github.com/ansible/ansible/pull/83391.

**Comment by bcoca** [comment]
next time i should add ci_complete to ensure nothing else breaks, this filter is clearly something used in many places

**Comment by s-hertel** [comment]
@mattclay I can move the test. It doesn't seem to be the only one relying on filters though, should I add an alias to the test target?

**Comment by mattclay** [comment]
@s-hertel No alias should be needed. Instead, we need to ensure that the `bool` filter has full test coverage within the `filter_core` integration test. Currently there are gaps in test coverage for the filter, some of which is made up by incidental testing in the `conditionals` integration test.

**Comment by s-hertel** [comment]
@mattclay The filter_core test had the coverage added for this case in the feature pull request (conflicting with the incidental coverage in `conditionals`). Since I've updated my PR to revert the feature though, now the filter_core test will again lack coverage.

**Comment by mattclay** [comment]
@s-hertel Could you open a PR with the new filter_core test, so we don't lose that?

**Comment by s-hertel** [comment]
@mattclay Sorry for my confusion, I just realized the coverage [did already exist](https://github.com/ansible/ansible/pull/83362/files#diff-e61d328c6629920ea251810f05fc0ea357ba74e5b62e7540606c7acd1dc64178L311), and was just rewritten. The conditionals incidental `conditionals` is covered intentionally also, and what happened here was the two just got out of sync.

**Pull request merged**